aten/src/ATen/core/jit_type.h

```diff
@@ -510,6 +511,94 @@ struct CAFFE2_API ListType : public SingleElementType<TypeKind::ListType, ListTy
  ListType(TypePtr elem) : SingleElementType(elem) {}
 };
 
+struct DictType;
+using DictTypePtr = std::shared_ptr<DictType>;
+struct DictType : public Type {
+  friend struct Type;
+  static const TypeKind Kind = TypeKind::DictType;
+
+  static DictTypePtr create(TypePtr key, TypePtr value) {
+    switch (key->kind()) {
+      case TypeKind::IntType:
+      case TypeKind::FloatType:
+      case TypeKind::StringType:
+        return DictTypePtr(new DictType(key, value));
+      default:
+        AT_ERROR(
+            "Cannot create dict for key type ",
+            key->str(),
+            ", only int, float, and string keys are supported");
+    }
+  }
+
+  std::string str() const override {
+    std::stringstream ss;
+    ss << getKeyType()->str() << "[" << getValueType()->str() << "]";
```

Comment:
Not sure it makes sense to have a different syntax, 'just because' for dictionaries. It is so close to array syntax `a[4] vs a[int]` that it will be confusing.